### PR TITLE
Fix account race condition (#13)

### LINKED ISSUE #13: Unlocking accounts race condition
Unlocking node accounts concurrently may result in transactions being blocked indefinitely

## Background

Currently the deployment and client code uses node accounts for signing transactions. This requires us to unlock the account before sending a transaction. We use [`web3::api::Personal::unlock_account`](https://docs.rs/web3/0.8.0/web3/api/struct.Personal.html#method.unlock_account) for that. We call this method without the timeout parameter. This means that the the next transaction for the account submitted with the `eth_sendTransaction` RPC method will be signed.

If the `eth_sendTransaction` method is called without the sender account being unlocked on the node the RPC method will not respond until the transaction is manually confirmed through the node API. Since we are not manually confirming transactions the RPC method effectively blocks.

## The Issue

The concrete issues arises if some code that unlocks an account and then sends a transaction is run concurrently. For example `oscoin_deploy::deploy` unlocks the dev account and then submits a contract creation transaction. If the end-to-end tests run two deploys in parallel then the following happens.

1. Thread 1 calls `unlock_account`.
2. Thread 2 calls `unlock_account`.
3. Thread 1 receives `unlock_account` response and sends the contract creation transaction.  
4. Thread 2 receives `unlock_account` response and sends the contract creation transaction.
5. The node receives the transaction from thread 1, process the transaction and locks the account again.
6. The node receives the transaction from thread 2. The account is locked and the node does not send a response.

This means that one of the threads will block indefinitely.

## Possible solutions

The easiest solution would be to add a timeout parameter to `web3::api::Personal::unlock_account` so that the account is unlocked for some time. The `web3` package does not seem to support this properly. Passing the parameter results in an error.

To solve the issue properly we should move away from `eth_sendTransaction` and use `eth_sendRawTransaction`. This would require us to replace some of the `web3` library code with our own code.

### PR DESCRIPTION
**Pull request by rockbmb**
Will close #13.

### CODE REVIEW

File: deploy/src/lib.rs [hunk 1]
```diff
@@ -45,21 +51,60 @@ pub fn deploy() -> Result<Contract<web3::transports::Http>, String> {
...
+        .expect("Transaction signing failed");
+
+    let sent_raw_tx = confirm::send_raw_transaction_with_confirmation(
+                eth_.transport().clone(),
+                signed_tx.raw,
+                poll_interval,
+                confirmations,
+        );
+
+    let pending_contract = PendingContract {
```
Comment (rockbmb) [inline deploy/src/lib.rs]: @geigerzaehler The fields of this constructor are private and it does not export a constructor method - I have absolutely no idea how to make this `pending_contract` work, I cannot come up with anything to produce a value of `PendingContract` here with just was `rust-web3` exports.

The only thing I can think of is submitting a PR to `rust-web3` that duplicates `execute` to use `send_raw_transaction_with_confirmation`. I'll get something ready and ask for your review.

**Comment by rockbmb** [comment]
CI fails with in the step

```
#!/bin/bash -eo pipefail
cargo build --package pwasm-utils-cli --bin wasm-build

wget https://releases.parity.io/ethereum/v2.6.0/x86_64-unknown-linux-gnu/parity
echo "1b50cabc8ce54983d1b10be4c4f5887ff4ecfe63177e6c49cde819a563fb9d96  parity" > parity.sum
sha256sum parity.sum
chmod +x parity
sudo mv parity /usr/local/bin
```

and I can't get build logs from CircleCI.

File: client/src/lib.rs [hunk 2]
```diff
@@ -248,25 +249,13 @@ impl Client {
...
-    /// Unlock the node account used by the client.
-    /// Note the `_` at the end to differentiate from the same function
-    /// imported from `web3`.
-    ///
-    /// TODO: Panics when the unlock RPC method responds with `false`. It should result in an
-    /// error.
-    fn unlock_account_(
+    /// Signs a transaction - this method is here to pass the empty string
+    /// password onto `rust-web3`'s `sign_transaction`.
+    fn sign_transaction_(
```
Comment (geigerzaehler) [inline client/src/lib.rs]: This method is tiny and only used in one place so it would be preferable to inline it.

**Comment by rockbmb** [comment]
@geigerzaehler I believe you changed the crate's dependencies in order for @xla's integration PR to work, but now it seems unable to fetch the correct version of `rust-web3`. `sign_and_execute` cannot be found, which I presume is because the wrong (i.e. not the latest) version of the crate is being built.

**Comment by rockbmb** [comment]
@geigerzaehler any other feedback?

**Pull request merged**